### DISCUSSION TITLE
Re-upload modified files

### DISCUSSION THREAD

**Discussion by KantarITCH** [abraunegg/onedrive#3025] [Q&A] [answered]
We are using this great tool to upload data to SharePoint. This is one direction only. We had to disable the upload validation (known issue). Is it still somehow possible to recognize if a file on SharePoint has been modified and in this case re-upload the file to keep consistency with the local data repository?

**Comment by abraunegg** [comment, marked as answer]
Not when using `--upload-only` as nothing online about the state of the file is retrieved.

**Reply by KantarITCH** [reply]
That's what I suspected. Would it be possible to implement something that would allow this?
I see that in this case you would need a record of the online data. And it would require the upload validation to work. 
I have read that the upload validation is based on the timestamp as the first option and an MD5 hash as the second option. My guess is that the upload validation is failing because the timestamp doesn't match. If there was a switch to select the preferred verification option (MD5 over timestamp) it would probably work.
Thank you very much for your work. I'm looking forward to your feedback!

**Reply by abraunegg** [reply]
> My guess is that the upload validation is failing because the timestamp doesn't match.

It fails because of a SharePoint *feature* .. more on that below.

> If there was a switch to select the preferred verification option (MD5 over timestamp) it would probably work.

It actually will not work to use the hash of the file as the first comparison.

Microsoft - when you use Business | Microsoft 365 | SharePoint ... all your 'OneDrive' data is stored on Microsoft SharePoint at the back end.

A 'feature' of Microsoft SharePoint is that post upload, SharePoint will *modify* your file by adding metadata to enrich it for SharePoint purposes.

There is zero capability to switch off this Microsoft SharePoint *feature* that that post upload, your data is not enriched.

When this enrichment happens, 2 things happen to the file online:
1. The hash of the file changes
2. The timestamp of the file changes

Effectively, the file you uploaded, is not the file you have locally anymore.

For further details, please read https://github.com/OneDrive/onedrive-api-docs/issues/935

**Reply by KantarITCH** [reply]
Ok, I see. Many thanks for the detailed explanation.
Then, the only option is to generate a hash after the upload and store it in a local database. Then it would have to be computed and compared with the stored hash regularly. When it doesn't match, it could be uploaded again.
I think this would be a feasible solution, but I'm not a software developer.

**Reply by abraunegg** [reply]
Not really - as the db hash needs to equal the local disk file and that is what is used to determine local change.

SharePoint complicates things - any upload - the file is different ... so the hash will always be different ... thus you will end up in an endless loop

**Reply by KantarITCH** [reply]
I have to contradict - or better elaborate my approach:
You would have to maintain a database which holds 3 values:
1. an ID for each local file
2. a hash for the local file
3. a hash for the remote file
whereby the two hashes don't have to match. 
To verify if the remote file has been modified, you would have to compute the hash of the remote file on a regular basis and then compare it with the 3rd value, stored in the database - not directly with the hash of the local file.

**Reply by abraunegg** [reply]
nope.

When OneDrive responds with the JSON data from ONLINE, post upload, the 'hash' data is this:

![image](https://github.com/user-attachments/assets/780833d5-bb12-4eb3-8692-1ba97ec7b0a0)

This hash, in a NON SharePoint situation matches the local file. When the JSON provided by Microsoft is saved, this hash is what is stored in the database. That hash is then used to compare the local on-disk file for changes.

The challenge you have is - the SharePoint feature - it *always* is changing the hash, thus, every compute is different, thus, always going to upload as your local file is *never* changing with the enrichment data.

Now ... the work being done to solve #3020 (effectively fix a regression and bring back how the client operated in v2.4.x) will in-part solve the SharePoint *feature* issue - however in an `--upload-only` scenario - the enriched file is never going to be pulled back down.

An option here would be to add a further configuration option that, *despite* `--upload-only` being used, when the integrity mis-match occurs, pull the online file back down to ensure everything is in-sync.

Is that a workable solution for you?

**Reply by KantarITCH** [reply]
This really sounds worse... and I've learned a lot. Thanks!
In fact I'm only interested to keep the integrity of the online file: If the online file is modified, I would like to re-upload the local file to restore the integrity of the online file.

**Reply by abraunegg** [reply]
> In fact I'm only interested to keep the integrity of the online file

However when using Microsoft OneDrive Business Accounts or a SharePoint Library - all which use  SharePoint at the back end - you CANNOT achieve this because of the Microsoft SharePoint feature (it is a bug, they call it a feature) that willfully modifies your file post upload.

Your file integrity is gone at that point.

**Reply by KantarITCH** [reply]
Ok, thank you very much for the exhaustive feedback. Very appreciated.
